### PR TITLE
Secret changes dialog added #274

### PR DESCRIPTION
**Pull request by harshith-1008**
## :mag: Overview

- This pull request solves issue #274.  
- Feature: Functionality which allows user to see the changes made to environment secrets. 

## :bulb: Proposed Changes

- Used GenericDialog component to show the changes made by user
- Added info button which onClick shows dialog 
- Created type SecretChange for the secret changes
- Created a component DisplayChanges which returns changes of a particular secret
- Also limited key value to 50 characters so that it does overflow out of the dialogue

## :framed_picture: Screenshots or Demo

Before
Screenshot 2024-07-31 at 6 53 37 PM

After
Screenshot 2024-07-31 at 7 34 42 PM
Screenshot 2024-07-31 at 7 34 55 PM

## :memo: Release Notes

User now can see the changes made to the environment secrets

## :question: Open Questions

Is there anyway I can enhance the UI more? 

### :test_tube: Testing

- Add a secret
- Change an attribute of a pre-existing secret
- Remove a tag from the secret and add other tags  

### :dart: Reviewer Focus

You can just make any desired changes and click on the info button to view the changes 

### :heavy_plus_sign: Additional Context

issue #274 

## :sparkles: How to Test the Changes Locally

- Pull the branch
- log in to your phase console account 
- head over to an environment of any app and start testing

### :green_heart: Did You...

- [x] Ensure linting passes (code style checks)?
- [ ] Update dependencies and lockfiles (if required)
- [ ] Regenerate graphql schema and types (if required)
- [ ] Verify the app builds locally?
- [x] Manually test the changes on different browsers/devices?

### CODE REVIEW

**Comment by harshith-1008** [comment]
@rohan-chaturvedi 
If possible, could you please look into this pr?

**Comment by rohan-chaturvedi** [comment]
@harshith-1008 Could you fix the filename please? 
image

**Comment by harshith-1008** [comment]
> @harshith-1008 Could you fix the filename please? image

Screenshot 2024-08-11 at 8 10 21 PM

@rohan-chaturvedi for some reason vscode isnt allowing me to keep this license file because folder name license already exists, can you please guide me to restore it?

**Comment by rohan-chaturvedi** [comment]
> > @harshith-1008 Could you fix the filename please? image
> 
> Screenshot 2024-08-11 at 8 10 21 PM
> @rohan-chaturvedi for some reason vscode isnt allowing me to keep this license file because folder name license already exists, can you please guide me to restore it?

Thanks a lot for the screenshot! Now I understand why this was a problem 😅 

I'll raise a PR to rename the license folder in a bit, that should resolve it.

**Comment by rohan-chaturvedi** [comment]
Hey @harshith-1008 , a couple of updates:
1) We've updated the UX for deleting secrets. Secrets can now be "staged" for delete, and are only deleted when clicking on "Deploy". See the PR for a full demo: https://github.com/phasehq/console/pull/340
This means we need to add Deleted secrets to the changed secrets dialog

2) I was working on a cleaner design for this dialog for while. I've made a [Figma mock here](https://www.figma.com/design/XGgL1d2uMi9PvbUKb2Ka2m/Undeployed-changed---detail-dialog?node-id=0-1&t=FDdZcgoFpbNyzNDi-1).

Here's a screenshot as well:
Frame 1(1)

You can grab the colors from Figma. The tailwind colors are:
- Dialog title: `text-zinc-900 dark:text-zinc-900`
- Dialog subtitle: `text-zinc-500`
- Created: `text-emerald-400` (Backgrounds: `bg-emerald-400/20 dark:bg-emerald-400/10`)
- Updated: `text-amber-400 ` (Backgrounds: `bg-amber-400/20 dark:bg-amber-400/10`)
- Deleted: `text-red-400` (Backgrounds: `bg-red-400/20 dark:bg-red-400/10`)

Sorry for the constant design updates! Lmk if you have any questions, we can hop on a slack huddle and discuss it as well.

**Comment by harshith-1008** [comment]
Screenshot 2024-08-15 at 4 29 31 PM

@rohan-chaturvedi I tried changing the UI, Can you guide me how to override text-color of tags? In Tag component the styling is (text-zinc-700 dark:text-zinc-200).

Screenshot 2024-08-15 at 4 31 44 PM

I tried this method but it isn't working. and should I decrease the height and width of tags?

**Comment by rohan-chaturvedi** [comment]
> Screenshot 2024-08-15 at 4 29 31 PM
> 
> @rohan-chaturvedi I tried changing the UI, Can you guide me how to override text-color of tags? In Tag component the styling is (text-zinc-700 dark:text-zinc-200).
> Screenshot 2024-08-15 at 4 31 44 PM
> 
> I tried this method but it isn't working. and should I decrease the height and width of tags?

@harshith-1008 You could add a `variant` prop to the Tag component that takes a string of type: `'normal' | 'deleted' | 'created'`. Apply the styles based on the value of this prop. Check out the Button component to see how a variant can be added. Make the `'normal'` default value of the prop, so that you won't have to add this prop wherever the Tag is currently used. 

Lets keep the height and width of the tags for now.

**Comment by harshith-1008** [comment]
@rohan-chaturvedi can you look into these changes? I have kept the tags width and height same. 

Screenshot 2024-08-16 at 10 41 51 AM

**Comment by rohan-chaturvedi** [comment]
> @rohan-chaturvedi can you look into these changes? I have kept the tags width and height same.
> 
> Screenshot 2024-08-16 at 10 41 51 AM

Looking good! Only issue I can see is that monospace font isn't used in the right places. Could you take a look at that?

**Comment by harshith-1008** [comment]
Screenshot 2024-08-16 at 3 45 32 PM

@rohan-chaturvedi  changed font irregularities.

**Comment by rohan-chaturvedi** [comment]
> Screenshot 2024-08-16 at 3 45 32 PM
> 
> @rohan-chaturvedi changed font irregularities.

* Use normal font for the change type ("Created", "Updated" etc)
* Use monospace for everything else (field names and values)

**Comment by harshith-1008** [comment]
> > Screenshot 2024-08-16 at 3 45 32 PM
> > @rohan-chaturvedi changed font irregularities.
> 
> * Use normal font for the change type ("Created", "Updated" etc)
> * Use monospace for everything else (field names and values)

Screenshot 2024-08-16 at 4 20 23 PM

Except change type, all other texts changed to font-mono.

**Comment by rohan-chaturvedi** [comment]
@harshith-1008 Nice! Lmk when the code changes are pushed, will take a look. Lets get this merged!

**Comment by harshith-1008** [comment]
@rohan-chaturvedi I have pushed changes.

File: frontend/app/[team]/apps/[app]/environments/[environment]/[[...path]]/page.tsx [hunk 1]
```diff
@@ -798,10 +968,66 @@ export default function EnvironmentPath({
...
+                            </div>
+                            <DisplayChanges change={change} />
+                          </div>
+                        )}
+                      </div>
+                    ))}
+                    {secretsToDelete?.map(secretId => {
+                      const deletedSecret = serverSecrets.find(secret => secret.id === secretId);
+                      return (
+                        <div className="flex items-center space-x-2 mb-6">
```
Comment (rohan-chaturvedi) [inline frontend/app/[team]/apps/[app]/environments/[environment]/[[...path]]/page.tsx]:
```suggestion
                        <div key={secretId} className="flex items-center space-x-2 mb-6">
```

Add a key for elements rendered in a list

File: frontend/app/[team]/apps/[app]/environments/[environment]/[[...path]]/page.tsx [hunk 2]
```diff
@@ -735,6 +755,156 @@ export default function EnvironmentPath({
...
+  } 
+  
+  const DisplayChanges = ({ change }: any) => {
+    const removedTags = change.tags?.old?.filter(tag => 
+      !change.tags?.new?.some(newTag => newTag.id === tag.id)
+    ) || [];
+  
+    const addedTags = change.tags?.new?.filter(tag => 
+      !change.tags?.old?.some(oldTag => oldTag.id === tag.id)
+    ) || [];
```
Comment (rohan-chaturvedi) [inline frontend/app/[team]/apps/[app]/environments/[environment]/[[...path]]/page.tsx]:
```suggestion
    const removedTags =
      change.tags?.old?.filter(
        (tag: SecretTagType) =>
          !change.tags?.new?.some((newTag: SecretTagType) => newTag.id === tag.id)
      ) || []

    const addedTags =
      change.tags?.new?.filter(
        (tag: SecretTagType) =>
          !change.tags?.old?.some((oldTag: SecretTagType) => oldTag.id === tag.id)
      ) || []
```

Add the `SecretTagType` annotation here

File: frontend/app/[team]/apps/[app]/environments/[environment]/[[...path]]/page.tsx [hunk 3]
```diff
@@ -798,10 +968,66 @@ export default function EnvironmentPath({
...
+                <span className="mr-[-0.5rem]" onClick={handleChangesReveal}>
+                  You have undeployed changes to this environment.
+                </span>
+                <GenericDialog
+                  buttonVariant=""
+                  title="Undeployed changes"
+                  onClose={() => setIsChangesModalOpen(false)}
+                  buttonContent={<FaInfoCircle />}
+                >
+                  <div className="flex flex-col space-y-2">
```
Comment (rohan-chaturvedi) [inline frontend/app/[team]/apps/[app]/environments/[environment]/[[...path]]/page.tsx]:
```suggestion
                  <div className="flex flex-col space-y-2 max-h-[85vh] overflow-auto">
```
Add a max height to handle overflow where there are a large number of changes

**Review by rohan-chaturvedi — CHANGES_REQUESTED** [review]
@harshith-1008 Awesome work! This is working perfectly :ok_hand: 

Just a few code changes and minor fixes required. Also, could you please create a separate component for this? Call is "DeployPreview" or something and put it in `components/environments/secrets`

**Comment by harshith-1008** [comment]
@rohan-chaturvedi I have pushed new changes. Can you look into it ?

**Review by rohan-chaturvedi — APPROVED** [review]

**Comment by rohan-chaturvedi** [comment]
Great work! Sorry for the length review :sweat_smile:

**Pull request merged**